ceno_rt/src/lib.rs

```diff
@@ -49,26 +50,40 @@ pub extern "C" fn sys_getenv(_name: *const u8) -> *const u8 {
 #[unsafe(no_mangle)]
 #[linkage = "weak"]
 pub unsafe extern "C" fn sys_rand(recv_buf: *mut u8, words: usize) {
-    unsafe fn step() -> u32 {
-        static mut X: u32 = 0xae569764;
-        // We are stealing Borland Delphi's random number generator.
-        // The random numbers here are only good enough to make eg
-        // HashMap work.
+    // We are stealing Borland Delphi's random number generator.
+    // The random numbers here are only good enough to make eg
+    // HashMap work.
+    fn step() -> u32 {
+        static X: AtomicU32 = AtomicU32::new(0xae569764);
+        let rslt = X.fetch_update(Ordering::Relaxed, Ordering::Relaxed, |x| {
+            Some(x.wrapping_mul(134775813).wrapping_add(1))
+        });
+        match rslt {
+            Ok(el) => el,
+            Err(err) => err,
+        }
+    }
+    let mut idx = 0;
+    let steps = words / 4;
+    let rest = words % 4;
+    for _ in 0..steps {
+        let [a, b, c, d] = step().to_le_bytes();
+        // SAFETY: Up to the caller
         unsafe {
-            X = X.wrapping_mul(134775813) + 1;
-            X
+            *recv_buf.add(idx) = a;
+            *recv_buf.add(idx.wrapping_add(1)) = b;
+            *recv_buf.add(idx.wrapping_add(2)) = c;
+            *recv_buf.add(idx.wrapping_add(3)) = d;
         }
+        idx = idx.wrapping_add(4);
     }
-    // TODO(Matthias): this is a bit inefficient,
-    // we could fill whole u32 words at a time.
-    // But it's just for testing.
```

Comment:
Fills 4 bytes at once. Shouldn't matter because the function is a mock but regardless 🤷